### ISSUE TITLE
Code review

### ISSUE THREAD

**Issue by samreid** [phetsims/balloons-and-static-electricity#201]
@jbphet and I will likely start instrumenting this simulation on Tuesday.  One of the preparatory steps is to perform a code review and bring it up to standards.

*NOTE! Prior to doing a code review, copy this checklist to a GitHub issue for the repository being reviewed.*

PhET code-review checklist
=============

**Build and Run Checks**

- [x] Does the sim build without warnings or errors?
- [x] Does the sim start up? (requirejs and built versions)
- [x] Does the sim experience any assertion failures? (run with query parameter 'ea')
- [x] Does the sim pass a scenery fuzzer test? (run with query parameter 'fuzzMouse&ea')

**Internationalization**
- [x] Are there any strings that are not being internationalized? (run with query parameter 'stringTest=x', you should see nothing but 'x' strings)
- [x] Does the sim layout gracefully handle internationalized strings that are twice as long as the English strings? (run with query parameter 'stringTest=double')
- [x] Does the sim layout gracefully handle internationalized strings that are exceptionally long? (run with query parameter 'stringTest=long')
- [x] Does the sim layout gracefully handle internationalized strings that are shorter than the English strings? (run with query parameter 'stringTest=X')
- [x] Does the sim stay on the sim page (doesn't redirect to an external page) when running with the query parameter 'stringTest=xss' (test passes if sim does not redirect, OK if sim crashes or fails to fully start). Only test on one desktop platform.
- [x] Make sure the string keys are all perfect, they are difficult to change after 1.0.0 is published.  Strings keys should
generally match the values, such as `{binaryProbability: "Binary Probability"}`. Screen names should use `screen.screenName` instead of
camelcase.  Message patterns and long paragraphs will also use a different pattern.

**Repository structure**

- [ ] Are all required files and directories present?

      For a sim repository named “my-repo”, the general structure should look like this (where assets/, audio/ or images/ may be omitted if the sim doesn’t have those types of assets).

```js
   my-repo/
      assets/
      audio/
         license.json
      doc/
         model.md
         implementation-notes.md
      images/
         license.json
      js/
         (see section below)
      .gitignore
      my-repo_en.html
      my-repo-strings_en.json
      Gruntfile.js
      LICENSE
      package.json
      README.md
```

   For a common-code repository, the structure is similar, but some of the files and directories may not be present if the repo doesn’t have audio, images, strings, or a demo application.

- [ ] Is the js/ directory properly structured?

      All JavaScript source should be in the js/ directory. There should be a subdirectory for each screen (this also applies for single-screen sims, where the subdirectory matches the repo name).  For a multi-screen sim, code shared by 2 or more screens should be in a js/common/ subdirectory. Model and view code should be in model/ and view/ subdirectories for each screen and common/.  For example, for a sim with screens “Introduction” and “Lab”, the general directory structure should look like this:

```js
   my-repo/
      js/
         common/
            model/
            view/
         introduction/
            model/
            view/
         lab/
            model/
            view/
         my-repo-config.js
         my-repo-main.js
         myRepo.js
 ```

- [ ] Is there a file in assets/ for every resource file in audio/ and images/? Note that there is *not necessarily* a
1:1 correspondence between asset and resource files; for example, several images may be in the same .ai file.
- [ ] Was the README.md generated by `grunt published-README` or `grunt unpublished-README`?
- [ ] Does package.json refer to any dependencies that are not used by the sim?
- [ ] Does sim-config.js refer to any dependencies that are not used by the sim?
- [ ] Is the LICENSE file correct? (GPL v3 for sims, MIT for common code)
- [ ] Does .gitignore match the one in simula-rasa?
- [ ] Are the credits in sim-main.js correct?
- [ ] Are there git repository branches that are no longer used and should be pruned?

**Coding conventions**

- [ ] Does the code style meet PhET's code-style guidelines, as described in the [“Coding Style Guidelines” section of "PhET Development Overview"](https://docs.google.com/document/d/1odXkliRagq0zuf1_NdOtQ2BrkC9hwlISnPi2y-dKdrk/edit#heading=h.1oxr3ptyo50w)?
- [ ] Is the code formatted according to PhET conventions? See [phet-idea-code-style.xml](https://github.com/phetsims/joist/blob/master/util/phet-idea-codestyle.xml) for IntelliJ IDEA code style.

**Documentation**

- [ ] Are documentation conventions followed, as described in the [“Coding Style Guidelines” section of the PhET Development Overview](https://docs.google.com/document/d/1odXkliRagq0zuf1_NdOtQ2BrkC9hwlISnPi2y-dKdrk/edit#heading=h.1oxr3ptyo50w)?
- [ ] Does model.md adequately describe the model, in terms appropriate for teachers?
- [ ] Does implementation-notes.md adequately describe the implementation, with an overview that will be useful to future maintainers?
- [ ] Are sim-specific query parameters (if any) identified and documented in one .js file? The .js file should be named <REPO>QueryParameters, for example ArithmeticQueryParameters for the aritmetic repository.

**Common Errors**

- [x] Is `Math.round` used where `dot.Util.roundSymmetric` should be used?  Math.round does not treat positive and negative numbers symmetrically, see https://github.com/phetsims/dot/issues/35#issuecomment-113587879
- [x] Is `toFixed` used where `dot.Util.toFixed` or `dot.Util.toFixedNumber` should be used? JavaScript's `toFixed` is notoriously buggy, behavior differs depending on browser, because the spec doesn't specify whether to round or floor.
- [x] Are random numbers using `phet.joist.random`, and all doing so after modules are declared (non-statically)?  For
instance, the following methods (and perhaps others) should not be used:
* `Math.random`
* `_.shuffle`
* `_.sample`
* `_.random`
* `new Random()`

**Organization, Readability, Maintainability**

- [ ] Does the organization and structure of the code make sense? Do the model and view contain types that you would expect (or guess!) by looking at the sim? Do the names of things correspond to the names that you see in the user interface?
- [ ] Are appropriate design patterns used?
- [ ] Is inheritance used where appropriate? Does the type hierarchy make sense?
- [ ] Is there any unnecessary coupling? (e.g., by passing large objects to constructors, or exposing unnecessary properties/functions)
- [ ] Is there too much unnecessary decoupling? (e.g. by passing all of the properties of an object independently instead of passing the object itself)?
- [ ] Are the source files reasonable in size? Scrutinize large files with too many responsibilities - can responsibilities be broken into smaller delegates?
- [ ] Are any significant chunks of code duplicated? This will be checked manually as well as with https://github.com/danielstjules/jsinspect
- [ ] Is there anything that should be generalized and migrated to common code?
- [ ] Are there any TODO or FIXME comments in the code?  They should be addressed or promoted to GitHub issues.
- [ ] Does the implementation rely on any specific constant values that are likely to change in the future? Identify constants that might be changed in the future. (Use your judgement about which constants are likely candidates.)
Does changing the values of these constants break the sim? For example, see https://github.com/phetsims/plinko-probability/issues/84.
- [ ] The simulation should use Property instead of PropertySet

**Performance, Usability**

- [ ] Does the sim perform as desired across the range of supported platforms? (eg, not too slow on slow platforms, not too fast on fast platforms)
- [ ] If the sim uses WebGL, does it have a fallback? Does the fallback perform reasonably well? (run with query parameter 'webgl=false')
- [ ] Are UI components sufficiently responsive? (especially continuous UI components, such as sliders)
- [ ] Are pointer areas optimized, especially for touch? (run with query parameter 'showPointerAreas')
- [ ] Do pointer areas overlap? (run with query parameter 'showPointerAreas')
- [ ] Is the timestep `dt` capped appropriately?  Try switching applications or browser tabs, then switch back.  Did the model take one big/long/awkward step forward?  If so, `dt` may need to be capped.  Example from `faradays-law.FaradaysLawModel`:

```js
// Cap large dt values, which can occur when the tab containing
// the sim had been hidden and then re-shown
dt = Math.min( 0.1, dt );
```

**Memory Leaks**

- [ ] Does a heap comparison using Chrome Developer Tools indicate a memory leak? (Describing this process is beyond the scope of this document.)
- [ ] For each common-code component (sun, scenery-phet, vegas, …) that opaquely registers observers or listeners, is there a call to that component’s dispose function, or documentation about why dispose is unnecessary?
- [ ] Are there leaks due to registering observers or listeners? These guidelines should be followed, or documentation added about why following them is not necessary:
	- [ ] AXON: `Property.link` is accompanied by `Property.unlink`.
	- [ ] AXON: `PropertySet.link` is accompanied by `PropertySet.unlink`.
	- [ ] AXON: Creation of `DerivedProperty` is accompanied by `dispose`.
	- [ ] AXON: Creation of `Multilink` is accompanied by `dispose`.
	- [ ] AXON: `Events.on` is accompanied by `Events.off`.
	- [ ] AXON: `Emitter.addListener` is accompanied by `Emitter.removeListener`.
	- [ ] SCENERY: `Node.on` is accompanied by `Node.off`
	- [ ] TANDEM: `tandem.addInstance` is accompanied by `tandem.removeInstance`.
- [ ] Do all types that require a `dispose` function have one? This should expose a public `dispose` function that calls
`this.disposeMyType()`, where `disposeMyType` is a private function declared in the constructor.  `MyType` should exactly
match the filename.

**PhET-iO**

- [ ] If the simulation is supposed to be instrumented for PhET-iO, please see [How to Instrument a PhET Simulation for PhET-iO]
(https://github.com/phetsims/phet-io/blob/master/doc/how-to-instrument-a-phet-simulation-for-phet-io.md)
for the PhET-iO development process.

**Comment by samreid** [comment]
Warnings during build:

```
~/github/balloons-and-static-electricity$ grunt
Running "eslint:allFiles" (eslint) task

Running "clean" task

Running "requirejs-build" task
>> license exception for BALLOONS_AND_STATIC_ELECTRICITY/charge-transfer-beep.mp3: This might NOT meet the license requirements for PhET, the use is temporary until something new is created
>> license exception for BALLOONS_AND_STATIC_ELECTRICITY/charge-transfer-beep.ogg: This might NOT meet the license requirements for PhET, the use is temporary until something new is created
>> license exception for BALLOONS_AND_STATIC_ELECTRICITY/bounds-beep.mp3: This might NOT meet the license requirements for PhET, the use is temporary until something new is created
>> license exception for BALLOONS_AND_STATIC_ELECTRICITY/bounds-beep.ogg: This might NOT meet the license requirements for PhET, the use is temporary until something new is created

Running "after-requirejs-build" task
>> Unused string: key=BALLOONS_AND_STATIC_ELECTRICITY/keyboardHelp.dialog, value=Balloon Hotkeys and Key Commands
```

**Comment by samreid** [comment]
> Make sure the string keys are all perfect

Skipping this since already published.

**Comment by samreid** [comment]
These files are missing:

      doc/
         model.md
         implementation-notes.md

**Comment by samreid** [comment]
I noticed there are many new files and types related to accessibility.  This means it may not be an exemplar for starting with a simple simulation for instrumentation with @jbphet.  On the other hand, it would be nice to understand how accessibility and phet-io can coexist.

**Comment by samreid** [comment]
On hold while discussing https://github.com/phetsims/balloons-and-static-electricity/issues/204#issuecomment-261024908

**Comment by samreid** [comment]
@jessegreenberg said he'll spend a bit more time tidying things up.  When it's ready, he'll assign this to me for review.

**Comment by jessegreenberg** [comment]
> Warnings during build:

> These files are missing:

These issues were addressed in other issues in this repo.

**Comment by jessegreenberg** [comment]
@samreid I believe this issue can be closed.  balloons-and-static-electricity was instrumented for PhET-iO, went through QA, and was published since this issue was opened.  Before an accessible version of BASE is published, it would be great to review the new features and investigate how they integrate with phet-io, but there is still work to be done then and I feel that a new review would be necessary at that time. Closing issue, but if you disagree, please feel free to reopen and assign back to me!

**Comment by samreid** [comment]
Sounds good, thanks!
